include/dca/math/nfft/dnfft_1d_gpu.hpp

```diff
@@ -0,0 +1,245 @@
+                  const int sign);
+
+  // Transforms the accumulated data to the frequency domain and stores it in f_w.
+  // If get_square is true, transforms the accumulated squared data instead.
+  template <typename OtherScalarType>
+  void finalize(func::function<std::complex<OtherScalarType>, func::dmn_variadic<WDmn, PDmn>>& f_w,
+                bool get_square = false);
+
+  // Sums the accumulated data in the time domain.
+  ThisType& operator+=(ThisType& other);
+
+  // Ensures that the arguments of the method 'accumulate' have been copied to the GPU.
+  void synchronizeCopy() {
+    m_copied_event_.block();
+  }
+
+private:
+  void initializeDeviceCoefficients();
+
+private:
+  using BaseClass::f_tau_;
+  static inline linalg::Vector<ScalarType, linalg::GPU>& get_device_cubic_coeff();
+
+  const double beta_;
+  cudaStream_t stream_;
+  const bool accumulate_m_sqr_;
+  linalg::Matrix<ScalarType, linalg::GPU> accumulation_matrix_;
+  linalg::Matrix<ScalarType, linalg::GPU> accumulation_matrix_sqr_;
+
+  linalg::Matrix<double, linalg::GPU> M_;
+  linalg::util::HostVector<details::ConfigElem> config_left_;
+  linalg::util::HostVector<details::ConfigElem> config_right_;
+  linalg::util::HostVector<ScalarType> times_;
+  linalg::Vector<details::ConfigElem, linalg::GPU> config_left_dev_;
+  linalg::Vector<details::ConfigElem, linalg::GPU> config_right_dev_;
+  linalg::Vector<ScalarType, linalg::GPU> times_dev_;
+
+  linalg::util::CudaEvent config_copied_event_;
+  linalg::util::CudaEvent m_copied_event_;
+};
+
+template <typename ScalarType, typename WDmn, typename RDmn, int oversampling>
+Dnfft1DGpu<ScalarType, WDmn, RDmn, oversampling, CUBIC>::Dnfft1DGpu(const double beta,
+                                                                    cudaStream_t stream,
+                                                                    const bool accumulate_m_sqr)
+    : BaseClass(), beta_(beta), stream_(stream), accumulate_m_sqr_(accumulate_m_sqr) {
+  initializeDeviceCoefficients();
+  assert(cudaPeekAtLastError() == cudaSuccess);
+  func::dmn_variadic<BDmn, BDmn, RDmn> bbr_dmn;
+  const int n_times = BaseClass::PaddedTimeDmn::dmn_size();
+  accumulation_matrix_.resizeNoCopy(std::make_pair(n_times, bbr_dmn.get_size()));
+  if (accumulate_m_sqr)
+    accumulation_matrix_sqr_.resizeNoCopy(accumulation_matrix_.size());
+}
+
+template <typename ScalarType, typename WDmn, typename RDmn, int oversampling>
+void Dnfft1DGpu<ScalarType, WDmn, RDmn, oversampling, CUBIC>::initialize() {
+  accumulation_matrix_.setToZero(stream_);
+  if (accumulate_m_sqr_)
+    accumulation_matrix_sqr_.setToZero(stream_);
+  assert(cudaPeekAtLastError() == cudaSuccess);
+}
+
+template <typename ScalarType, typename WDmn, typename RDmn, int oversampling>
+void Dnfft1DGpu<ScalarType, WDmn, RDmn, oversampling, CUBIC>::initializeDeviceCoefficients() {
+  static bool initialized = false;
+  static std::mutex mutex;
+  if (initialized)
```

Comment:
It seems that "initialized" is always false at this point.  What is the purpose of this conditional check?

Reply:
Prevent multiple initializations. It might be called by multiple threads multiple time. The first initialized check avoids locking and unlocking the mutex on subsequent iterations.
A cleaner implementation can be achieved with std::call_once, but I was not aware of its existence. I might change it later if I have the time.

Reply:
Implemented in 1caf9406fcc8c5a1e607ae59764749e88572df10.

Reply:
I see.  The new implementation is clearer in what the code is doing.